pyro/contrib/epidemiology/util.py

```diff
@@ -8,54 +8,152 @@
+            8723088 - 1629936 * ss + 110880.0 * s4 - 2772 * s6,
+            8723088 - 1629936 * tt + 110880.0 * t4 - 2772 * t6,
+            3800016 + 3503808 * s + 365904 * ss - 443520 * s3 - 55440 * s4 + 33264 * s5 - 2772 * s6,
+            1017423 + 1823283 * s + 1058211 * ss + 51975 * s3 - 148995 * s4 - 18711 * s5 + 20097 * s6 - 3267 * s7,
+            84744 + 282744 * s + 382536 * ss + 249480 * s3 + 55440 * s4 - 24948 * s5 - 18018 * s6 + 5445 * s7,
+            693 + 4851 * s + 14553 * ss + 24255 * s3 + 24255 * s4 + 14553 * s5 + 4851 * s6 - 3267 * s7,
+            693 * s7,
+        ], dim=-1) * (1/32931360)
+    elif num_quant_bins == 16:
+        global w16
+        if w16.device != s.device:
+            w16 = w16.to(s.device)
+        s_powers = s.unsqueeze(-1).unsqueeze(-1).pow(torch.arange(10.))
+        t_powers = t.unsqueeze(-1).unsqueeze(-1).pow(torch.arange(10.))
+        splines_t = (w16 * t_powers).sum(-1)
+        splines_s = (w16 * s_powers).sum(-1)
+        index = [0, 1, 2, 3, 4, 5, 6, 15, 7, 14, 13, 12, 11, 10, 9, 8]
+        probs = torch.cat([splines_t, splines_s], dim=-1)
+        probs = probs.index_select(-1, torch.tensor(index))
+
+    return probs
+
+
+def quantize(name, x_real, min, max, num_quant_bins=4):
     """Randomly quantize in a way that preserves probability mass."""
     assert min < max
     lb = x_real.detach().floor()
 
-    # This cubic spline interpolates over the nearest four integers, ensuring
-    # piecewise quadratic gradients.
-    s = x_real - lb
-    ss = s * s
-    t = 1 - s
-    tt = t * t
-    probs = torch.stack([
-        t * tt,
-        4 + ss * (3 * s - 6),
-        4 + tt * (3 * t - 6),
-        s * ss,
-    ], dim=-1) * (1/6)
+    probs = compute_bin_probs(x_real - lb, num_quant_bins=num_quant_bins)
+
     q = pyro.sample("Q_" + name, dist.Categorical(probs),
                     infer={"enumerate": "parallel"})
-    q = q.type_as(x_real) - 1
+    q = q.type_as(x_real) - (num_quant_bins // 2 - 1)
 
     x = lb + q
     x = torch.max(x, 2 * min - 1 - x)
     x = torch.min(x, 2 * max + 1 - x)
+
     return pyro.deterministic(name, x)
 
 
-def quantize_enumerate(x_real, min, max):
+def quantize_enumerate(x_real, min, max, num_quant_bins=4):
     """Quantize, then manually enumerate."""
     assert min < max
     lb = x_real.detach().floor()
 
-    # This cubic spline interpolates over the nearest four integers, ensuring
-    # piecewise quadratic gradients.
-    s = x_real - lb
-    ss = s * s
-    t = 1 - s
-    tt = t * t
-    probs = torch.stack([
-        t * tt,
-        4 + ss * (3 * s - 6),
-        4 + tt * (3 * t - 6),
-        s * ss,
-    ], dim=-1) * (1/6)
+    probs = compute_bin_probs(x_real - lb, num_quant_bins=num_quant_bins)
     logits = safe_log(probs)
-    q = torch.arange(-1., 3.)
+
+    arange_min = - (num_quant_bins // 2 - 1)
+    arange_max = num_quant_bins // 2 + 1
```

Comment:
nitnit: I guess you could write these as
```py
    arange_min = 1 - num_quant_bins // 2
    arange_max = 1 + num_quant_bins // 2
```

Reply:
done